File: Solution/623.在二叉树中增加一行/623.在二叉树中增加一行.java
/*
 * @Descroption: LeetCode 623. 在二叉树中增加一行
 * @Author: EmoryHuang
 * @Date: 2022-08-04 09:08:05
 * @Method:
 * 层序遍历，遍历到 depth - 1 层时，添加新的节点
 */

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

  public TreeNode addOneRow(TreeNode root, int val, int depth) {
    if (depth == 1) {
      TreeNode newNode = new TreeNode(val);
      newNode.left = root;
      return newNode;
    }
    Queue<TreeNode> q = new LinkedList<>();
    q.offer(root);
    int cur_depth = 1;
    while (!q.isEmpty()) {
      int sz = q.size();
      for (int i = 0; i < sz; i++) {
        TreeNode node = q.poll();
        if (cur_depth == depth - 1) {
          TreeNode newNode = new TreeNode(val);
          newNode.left = node.left;
          node.left = newNode;
          newNode = new TreeNode(val);
          newNode.right = node.right;
          node.right = newNode;
        }
        if (node.left != null) q.offer(node.left);
        if (node.right != null) q.offer(node.right);
      }
      cur_depth++;
    }
    return root;
  }
}

File: Solution/623.在二叉树中增加一行/623.在二叉树中增加一行.cpp
/*
 * @Descroption: LeetCode 623. 在二叉树中增加一行
 * @Author: EmoryHuang
 * @Date: 2021-08-16 19:44:55
 * @Method:
 * 层序遍历，遍历到 depth - 1 层时，添加新的节点
 */

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
   public:
    TreeNode* addOneRow(TreeNode* root, int val, int depth) {
        if (depth == 1) {
            TreeNode* node = new TreeNode(val);
            node->left = root;
            return node;
        }
        queue<TreeNode*> q;
        q.push(root);
        int cur_depth = 1;
        while (!q.empty()) {
            int n = q.size();
            for (int i = 0; i < n; i++) {
                auto node = q.front();
                q.pop();
                if (cur_depth == depth - 1) {
                    TreeNode* tmpl = node->left;
                    node->left = new TreeNode(val);
                    node->left->left = tmpl;

                    TreeNode* tmpr = node->right;
                    node->right = new TreeNode(val);
                    node->right->right = tmpr;
                }
                if (node->left) q.push(node->left);
                if (node->right) q.push(node->right);
            }
            cur_depth++;
        }
        return root;
    }
};